File: leetcode-python/String/LC0003-无重复字符的最长子串.py
'''
3. 无重复字符的最长子串
输入: s = "abcabcbb"
输出: 3
解释: 因为无重复字符的最长子串是 "abc"，所以其长度为 3
'''


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        d, ans, l = {}, 0, -1
        for r in range(len(s)):
            if s[r] in d:
                l = max(d[s[r]], l)
            d[s[r]] = r
            ans = max(ans, r - l)
        return ans

File: leetcode-go/String/LC0003-无重复字符的最长子串.go
package main

func lengthOfLongestSubstring(s string) int {
	m := make(map[byte]int)
	n := len(s)
	k, ans := -1, 0
	for i := 0; i < n; i++ {
		if i != 0 {
			delete(m, s[i-1])
		}
		for k+1 < n && m[s[k+1]] == 0 {
			m[s[k+1]]++
			k++
		}

		ans = max(ans, k-i+1)
	}
	return ans
}

func max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

func main() {
	s := "ababcbb"
	lengthOfLongestSubstring(s)
}
